library/common/engine.h

```diff
@@ -38,9 +38,12 @@ class Engine {
   /**
    * Increment a counter with a given string of elements and by the given count.
    * @param elements, joined elements of the timeseries.
+   * @param tag_arr, the array of {key, value} pairs of tags.
+   * @param tag_arr_len, length of tag_arr
    * @param count, amount to add to the counter.
    */
-  envoy_status_t recordCounterInc(const std::string& elements, uint64_t count);
+  envoy_status_t recordCounterInc(const std::string& elements, const char* tag_arr[][2],
```

Comment:
use a 3d array to pass the tag key value pairs, it's the only structure compilable for c, cpp and objective-c

Reply:
Can we somehow leverage the bridging we do for headers, which is also a set of key, value pairs?

Reply:
i read through code on headers, it brings extra complexity to leverage the bridge. 

there is [another comment](https://github.com/envoyproxy/envoy-mobile/pull/1271#discussion_r586909893) on creating a struct, that sounds a good idea to me.

Reply:
The other comment was my initial thought before I realized the tags were exactly the same data structure as `envoy_headers`. 

I don't think reusing that data structure brings additional complexity. In fact it might reduce some given that we can re-use utility functions. 

My thought is that we can rename `envoy_headers` to `envoy_repeated_string_pair` (or something better); changing the field names to approriate names. Then `typedef envoy_repeated_string_pair envoy_headers` and same with `envoy_pulse_tags`.

The only difference I see here is in memory management, where with `envoy_headers` we have allocated memory that envoy mobile's c/c++ manages, but with your PR it seems that `char *` (with memory owned by the platform) is passed down.

As a reminder, I am out the next two days, but @goaway can advise.

Reply:
i get the inspiration for re-use, in general it's a good idea.

i looked more closely about the renaming, `envoy_header` carries the `context` and `envoy_release_f` fields ([link](https://github.com/envoyproxy/envoy-mobile/blob/43105fd9b89a7e9d4f6e723a8d47e7ccbaee0251/library/common/types/c_types.h#L85-L86)) from `envoy_data`, they are essential for headers, they are not necessary for tags. I'm aware that one option is split out a standalone `envoy_repeated_string_pair` struct and make another level of embedding for `envoy_header`, i'm not seeing the gain for the split/refactor at this point.

i'd like to hear more about the differences you brought in on memory management, what are the risks you see of having the memory allocated on the envoy mobile c/c++ vs platform? i guess the most obvious answer is memory leak, and i'd like to hear more about the intricacy of it.

Reply:
Let me give more context. At a high level, the header data structure and the tags data structure is the same. It is an array of string pairs. 

Envoy Mobile has comprehensive infrastructure (types, utility functions, tests) for bridging an array of string pairs across all supported languages. Of course the more specific application is for headers. My high level point is that we should leverage the same infrastructure for another specific application of an array of string pairs; tags.

It is true that the memory management between tags and headers is different based on the code you have in this PR, but that is part of the reason I am advocating for leveraging the already existing infrastructure: if we have all arrays of string pairs (headers, tags) using the same infrastructure, any improvements we make will be used by both. To give a concrete example: https://github.com/envoyproxy/envoy-mobile/issues/949 will fully bridge the types we are discussing here allowing less copy overhead. If we use the same underlying types for both headers and tags, that improvement will be immediately available to both.

Lastly, thanks to `typedef` we will be able to alias all necessary types such that readability is not hindered, and in fact, improved over passing individual pointers.